File: Baekjoon/Python/DFS와 BFS.py
from collections import deque


def bfs(graph, visited_bfs, start):
  dq = deque([start])
  visited_bfs[start] = True
  while dq:
    cur = dq.popleft()
    print(cur, end=' ')
    for next in graph[cur]:
      if not visited_bfs[next]:
        dq.append(next)
        visited_bfs[next] = True


def dfs(graph, visited_dfs, cur):
  visited_dfs[cur] = True
  print(cur, end=' ')
  for next in graph[cur]:
    if not visited_dfs[next]:
      dfs(graph, visited_dfs, next)


if __name__ == '__main__':
  n, m, v = map(int, input().split())
  graph = [[] for _ in range(n + 1)]
  visited_bfs = [False] * (n + 1)
  visited_dfs = [False] * (n + 1)

  for i in range(m):
    start, end = map(int, input().split())
    graph[start].append(end)
    graph[end].append(start)

  for i in range(n+1):
    graph[i].sort()

  dfs(graph, visited_dfs, v)
  print()
  bfs(graph, visited_bfs, v)

File: Baekjoon/java/DFS와 BFS.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

class Main {
    static boolean[] isVisited;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int v = Integer.parseInt(st.nextToken());
        
        isVisited = new boolean[n + 1];
        List<Integer>[] edges = new ArrayList[n + 1];
        for(int i = 0; i <= n; i++) {
            edges[i] = new ArrayList<>();
        }
        
        for(int i = 0; i < m; i++) {
            StringTokenizer st2 = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st2.nextToken());
            int end = Integer.parseInt(st2.nextToken());
            edges[start].add(end);
            edges[end].add(start);
        }
        
        for(int i = 0; i <= n; i++) {
            Collections.sort(edges[i]);
        }
        
        isVisited[v] = true;
        dfs(edges, v);
        
        Arrays.fill(isVisited, false);
        System.out.printf("\n");
        
        bfs(edges, v);
    }
    
    public static void bfs(List<Integer>[] edges, int start) {
        Queue<Integer> queue = new LinkedList<>();
        
        queue.add(start);
        isVisited[start] = true;
        while(!queue.isEmpty()) {
            int cur = queue.poll();
            System.out.printf("%d ", cur);
            
            List<Integer> nextNodes = edges[cur];
            for(int next : nextNodes) {
                if(isVisited[next]) continue;
                isVisited[next] = true;
                queue.add(next);
            }
        }
    }
    
    public static void dfs(List<Integer>[] edges, int cur) {
        System.out.printf("%d ", cur);
        
        List<Integer> nextNodes = edges[cur];
        for(int next : nextNodes) {
            if(isVisited[next]) continue;
            isVisited[next] = true;
            dfs(edges, next);
        }
    }
}
